File: LC/[0001] Two Sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> val2idx;

        for(int ii=0; ii<nums.size(); ++ii){
            auto curr=nums[ii];
            if(val2idx.find(target-curr) != val2idx.end()){
                return {ii, val2idx[target-curr]};
            }
            val2idx[curr]=ii;
        }
        return {};
    }
};

File: LC/[0001] Two Sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        : all combination N^2
        : use a map to cache all that is visited for complement lookup
        """

        val2idx={}

        for ii, vv in enumerate(nums):
            if target-nums[ii] in val2idx:
                return (val2idx[target-nums[ii]], ii)
            else:
                val2idx[vv]=ii


        return None # shouldn't be here
